### ISSUE TITLE
become: pbrun doesn't allow to execute multiple commands

### ISSUE THREAD

**Issue by bushvin** [ansible/ansible#12167]
pbrun doesn't allow to run multiple commands by joining them with a **;** or **&&**. ...

Ansible v1.9.2
pbrun v7.0.1-02

Hence the command generated by ansible to launch python remotely, together with echo BECOME-SUCCESSFULL fails at every attempt.

The culprit seems to be line **1256** in **lib/ansible/utils/__ init __.py**:

```
becomecmd = '%s -b %s -u %s "%s"' % (exe, flags, user, pipes.quote('echo %s; %s' % (success_key,cmd)))
```

I have no idea why to run the echo with elevated rights (maybe someone could explain it to me) but when changing this line to 

```
becomecmd = 'echo %s; %s -b -l %s -u %s %s' % (success_key, exe, flags, user, cmd)
```

seems to resolve the issue.

**Comment by bcoca** [comment]
In my tests it worked with pbrun 8.5.1, if it is not a version issue it probably depends on the profile, I use a very permissive one (basically all as root).

The echo is there INSIDE the privilege escalation so we can detect that the escalation itself worked, moving it outside would give us a "success" even when it fails.

**Comment by bcoca** [comment]
@bushvin so I cannot get an older version to test against, is there any way you can change the profiles to allow for multiple commands run separated by ;? will it work if we && them?

**Comment by bushvin** [comment]
@bcoca It makes sense to put the echo within the elevation command. I hadn't thought of that.

I'll try to figure out how to set the "Multiple commands" option, but if you have any leads, those would be more than welcome, as I have kind of inherited this mess without any handover, and I'm stuck with it, as it's engrained within the DNA of the company I work for.

I can't promise to get you any results tomorrow, but I'll definitely be able to test it wednesday.

**Comment by bcoca** [comment]
I'm not a powerbroker expert, I just got a demo version and the simplest config i could to test the execution. I really don't know where you would set/unset a multiple command restriction.

**Comment by bushvin** [comment]
Fair enough, I'll see what I can find.

**Comment by bushvin** [comment]
Hi Brian,

I discovered the ";" sign is considered a threat in our powerbroker policy,
so the way ansible invokes pbrun is correct.
Not a bug. You can close this thread.

For posterity: the powerbroker policy is called pb.conf, there's no default
location, as everything depends on the initial setup.

Thanks for your help!

On Mon, 31 Aug 2015 21:14 Brian Coca notifications@github.com wrote:

> I'm not a powerbroker expert, I just got a demo version and the simplest
> config i could to test the execution. I really don't know where you would
> set/unset a multiple command restriction.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible/issues/12167#issuecomment-136472010.

## 

Disclaimer http://www.critter.be/maildisclaimer

**Comment by bcoca** [comment]
@bushvin thanks for posting the answer! closing this ticket then (you as owner can also close).

**Comment by bushvin** [comment]
That's an email feature a I have yet to discover...

On Wed, 2 Sep 2015 18:14 Brian Coca notifications@github.com wrote:

> Closed #12167 https://github.com/ansible/ansible/issues/12167.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible/issues/12167#event-399130257.

## 

Disclaimer http://www.critter.be/maildisclaimer

**Comment by randyr505** [comment]
We are running an old version of pbrun (6.2.0-09) and here is what I had to do with ansible to get it working:
latest ansible 2.0 git source (as of 12/31/15)
make clean
changed ansible-2.0/lib/ansible/playbook/play_context.py as follows:
from
     becomecmd = '%s -b %s -u %s %s' % (exe, flags, self.become_user, success_cmd)
to
      cmd = cmd.replace("LANG=en_US.UTF-8 ", "")
      cmd = cmd.replace("LC_MESSAGES=en_US.UTF-8 ", "")
      cmd = cmd.replace("LC_ALL=en_US.UTF-8 ", "")
      becomecmd = 'echo %s; %s -b -l %s -u %s %s' % (success_key, exe, flags, self.become_user, cmd)
I realize this may not be the best method in python for the replace (and I am new to python) but it works.
make
